### PR TITLE
Refactored retrieving of remote projects and added support for FTP, Mercurial and Subversion

### PR DESCRIPTION
**Pull request by saserr**

### CODE REVIEW

**Comment by saserr** [comment]
@harrah Should I now inline these simple methods like `retrieveRODir` and `downloadAndExtract`?

@vigdorchik Can you check new code?

**Comment by harrah** [comment]
@saserr If they are only used in one place, that is fine. Otherwise, leave them.  Also, you don't need to rebase anymore, just add new commits.  It seems to be the better process.

**Comment by vigdorchik** [comment]
I think you should have 2 'run' methods:

def run(command: String*): Unit = run(command, cwd = None)

and

def run(command: Seq[String], cwd: Option[File])

That way most calls won't carry the ugly 'None' in the front of argument list.

**Comment by saserr** [comment]
Ok, I'll add another run method later this day. Thanks for help!

**Comment by vigdorchik** [comment]
Another, probably stupid, question. What happens if there are changes on remote between the 1st and the 2nd calls to the resolver? The changes are not pulled in, are they?

**Comment by saserr** [comment]
Updating of remote projects is still not supported in sbt, and it is on wishlist (number 3 on https://github.com/harrah/xsbt/wiki/Opportunities). I was thinking on implementing it when i finish with this pull request. But, i will probably need some guidance with that implementation.

**Comment by saserr** [comment]
@vigdorchik i fixed `!file.exists` and added second `run` method. can you check the code now?

**Comment by vigdorchik** [comment]
Also, if you make RichURI into case class, the copy method will be automatically provided.

**Comment by saserr** [comment]
It won't work, because i need the URI as parameter in constructor, because I misuse it in `git:http:...` like URIs.

**Comment by vigdorchik** [comment]
Sorry, I'm not following. Can you please elaborate more?

**Comment by saserr** [comment]
Because, for example mercurial has no mercurial-specific things in it's URI, so it is impossible to distinguish it from just plain 'http' URI. So, using idea from maven i added these marker schemes like 'hg:http:...' which distinguish different URIs. Downside of this is that only getScheme and getSchemeSpecificPart methods will return not-null values and some will even throw MalformedURIException.

So if i make RichURI a case class, I would have to define all properties of URI in constructor by myself and modify the implicit fromURI method in the RichURI's companion object to use this new constructor. Now we come to problem where for URIs that have marker scheme, I can't create RichURI because it might throw MalformedURIException. I could go around it by creating another property called markerScheme, but that would make code much more complicated and longer.

Another reason to not do it is that either I define all properties of URI in clone method or in the constructor + the implicit fromURI method which does not change the number of lines of code.

**Comment by vigdorchik** [comment]
I see. Case class was indeed a bad suggestion. Looks good to me once you make the last commit we discussed.

**Comment by saserr** [comment]
But what about when URI is not a directory thing?

**Comment by vigdorchik** [comment]
I don't feel strongly about it.

**Comment by saserr** [comment]
@vigdorchik i tested the code manually today and found few bugs. i fixed them all, but one was really tough so i had to change how cloning works for git in 96f68fb. can you please review the changes and see if they are ok?

**Comment by vigdorchik** [comment]
I'll try to do it tomorrow.

**Comment by saserr** [comment]
@vigdorchik thanks

**Comment by harrah** [comment]
- Don't arbitrarily introduce a symbolic method.  A normal method toStdout(...) would be fine.
- However, output should go to a logger and not standard output.
- Calling lines and then printing it converts the byte stream to a Stream back to a byte stream.  Just redirect the output to a logger.

As for the two-step procedure, I think it is worth the benefit unless there are fundamental problems.  If the current issues are minor things to work through, then let's just work through them.  If the code needs to be made more understandable, let's do that.

@saserr, I can see how it is likely to save time and space, but I don't know the exact details.  Can you be precise about what git does with respect to network access and hardlinking files?  What about on Windows? (does hardlinking exist there?)

@vigdorchik can you summarize your problems with the two-step process?

**Comment by vigdorchik** [comment]
@harrah My point is that the additional caching step introduces new bug possibilities. Right now I only see the problem with a new branch created on remote, which is not tracked on local. Maybe there are other problems too. So in the least the current code should account for new branches.

**Comment by saserr** [comment]
@harrah idea is very simple. before, when two projects required two different branches of same repository, the code would clone the remote repository twice. new idea is to have a clean copy of repository locally for HEAD branch and clone from it, thus reducing number of remote clonings. first, copying data over network and on filesystem is order of magnitude faster. furthermore, both git and mercurial use hardlinks for immutable data (basically almost everything under .git and .hg) when local cloning is used. only fat32 does not support hardlinks and there both of them fallback to pure copying, but still this is much faster then getting data from remote repository.

@vigdorchik i might have another solution that would not require code to manually track remote branches. i'll post update today or tomorrow.

**Comment by harrah** [comment]
I think the two step will be worth it, but I see @vigdorchik's point that there are some things to think about.  I think we agree the other parts are not controversial and should go in for 0.12.0.  Then, @saserr should split the two-step pieces out into a separate pull request and the other parts can get merged now.

**Comment by saserr** [comment]
Sorry for the delay. I removed the optimization and tested the code. Everything works fine and I think that it can be merged in.
I will continue to work on the optimization + I want to added logging into the code so user have much better feedback on what sbt is doing. I will create a new pull request as soon as I have working code.

**Comment by harrah** [comment]
@vigdorchik, ok to merge now?

**Comment by vigdorchik** [comment]
@saserr @harrah I'm afraid there are more problems, namely reload doesn't recompile if the build definition file changed. This can be reproduced both trying to test manually and with 'sbt "scripted actions/reload"'. I didn't examine the exact source of the problem.

**Comment by saserr** [comment]
@vigdorchik I don't understand your problem. How do you reproduce the error?

**Comment by vigdorchik** [comment]
@saserr publish-local your version, then in a test project try 'reload' after making a change to .sbt file. Nothing is recompiled.

**Comment by harrah** [comment]
The problem is that a writeable local directory is always copied now instead of the previous behavior that used the project right out of the original directory.  Source dependencies aren't updated automatically, so the change isn't seen.

I've merged the commits and made this fix.

**Comment by saserr** [comment]
Thanks for all the help and patience with me.

@harrah How should updating of remote projects work? Should resolver just try to update the code when it sees that there is already a directory in staging for that project?
